src/Stratis.Bitcoin/P2P/Peer/NetworkPeer.cs

```diff
@@ -544,10 +546,13 @@ private async Task ProcessMessageAsync(IncomingMessage message, CancellationToke
             }
             finally
             {
-                if (this.onDisconnectedAsyncContext.Value.DisconnectCallbackRequested)
-                    this.onDisconnected(this);
+                if (!insideCallback)
+                {
+                    if (this.onDisconnectedAsyncContext.Value.DisconnectCallbackRequested)
+                        this.onDisconnected(this);
 
-                this.onDisconnectedAsyncContext.Value = null;
+                    this.onDisconnectedAsyncContext.Value = null;
```

Comment:
This will reset the AsuncLocal for the next message to be processed.

Reply:
See previous comment.